drivers/usb/device/usb_dc_stm32.c

```diff
@@ -778,12 +778,21 @@ int usb_dc_ep_mps(const u8_t ep)
 
 void HAL_PCD_ResetCallback(PCD_HandleTypeDef *hpcd)
 {
+	int i;
+
 	SYS_LOG_DBG("");
 
 	HAL_PCD_EP_Open(&usb_dc_stm32_state.pcd, EP0_IN, EP0_MPS, EP_TYPE_CTRL);
 	HAL_PCD_EP_Open(&usb_dc_stm32_state.pcd, EP0_OUT, EP0_MPS,
 			EP_TYPE_CTRL);
 
+	/* The DataInCallback will never be called at this point for any pending
+	 * transactions.
+	 * */
+	for (i = 0; i < CONFIG_USB_NUM_BIDIR_ENDPOINTS; i++) {
+		k_sem_init(&usb_dc_stm32_state.in_ep_state[i].write_sem, 1, 1);
+	}
```

Comment:
I agree that we should reset semaphores during a reset. Do we need to initialize them again? Did you try to reset them with `k_sem_reset()` instead?

The comment also is confusing. Something like "Reset IN endpoints semaphores" sounds better to me.

Reply:
Agree k_sem_reset seems to be the right way, avoiding unbalanced sem after reset.

Reply:
@atx , reason not to follow @loicpoulain & @ydamigos request?

Reply:
@erwango See my [earlier comment](https://github.com/zephyrproject-rtos/zephyr/pull/9412#issuecomment-424989596)

The semaphores have maximum permitted count of 1 and k_sem_give will therefore either set them to 1 or keep them at 1. Furthermore, it's unclear to me if k_sem_reset is ISR-safe as the documentation does not explicitly say so.